### PR TITLE
Make `generate_ext4_image` tool deterministic

### PR DESCRIPTION
**Pull request by bduffany**
* Set `mkfs.ext4` flags to ensure determinism (see comments in `ext4.go`)
* Fix log noise by running the tool with `--app.log_level=warn`

### CODE REVIEW

File: enterprise/server/util/ext4/ext4.go [hunk 1]
```diff
@@ -38,6 +40,15 @@ const (
...
 	debugfsPath = "/sbin/debugfs"
+
+	// Filesystem UUID and directory hash seed used for reproducible images.
+	// mke2fs otherwise generates both randomly.
+	reproducibleUUID     = "3b3f1e6a-2a0c-4d1e-9d7e-6f0a7c1b2d3e"
+	reproducibleHashSeed = "9c8b7a6d-5e4f-4a3b-8c2d-1e0f9a8b7c6d"
+)
+
+var (
+	reproducibleImages = flag.Bool("executor.reproducible_ext4_images", false, "If true, fix the filesystem UUID and directory hash seed of ext4 images created from directories, skip copying extended attributes, and clamp all timestamps to the newest file mtime in the tree, so that a freshly extracted directory tree produces a byte-identical image on every run. Requires e2fsprogs 1.47.2 or later.", flag.Internal)
```
Comment (dan-stowell) [inline enterprise/server/util/ext4/ext4.go]: Why gate this behind a flag? So we only change behavior for tests and don't accidentally impact anyone else?

Comment (bduffany) [reply enterprise/server/util/ext4/ext4.go]: yeah, the timestamp/xattrs changes are especially risky - don't want to unintentionally break people with those

File: enterprise/server/util/ext4/ext4.go [hunk 2]
```diff
@@ -163,12 +217,19 @@ func DiskSizeBytes(ctx context.Context, inputDir string) (int64, error) {
 		if err != nil {
 			return err
 		}
-		// stat() does not account for file or symlink metadata or for
-		// filesystem data structures like indirect blocks which consume disk
-		// space, so add 2 extra disk blocks for each entry as a rough way to
-		// account for this. Also note that stat() blocks are always 512 bytes
-		// regardless of the FS settings.
-		total += blockSize + info.Sys().(*syscall.Stat_t).Blocks*512
+		// Estimate from the entry type and file size rather than from the
```
Comment (dan-stowell) [inline enterprise/server/util/ext4/ext4.go]: Does this mean the image could take up more space on disk than indicated by `DiskSizeBytes`? Can that bite us anywhere?

Comment (bduffany) [reply enterprise/server/util/ext4/ext4.go]: It can in theory bite us, but should be pretty unlikely at this point - we multiply the estimate by 1.2 as a safeguard, and we also add a fixed lowerbound of 4MB. We also test against a pretty decent variety of images at this point

**Review by dan-stowell — APPROVED** [review]

**Review by tylerwilliams — APPROVED** [review]
would love to flip this on and remove the flag if it works ok

File: enterprise/server/util/ext4/ext4.go [hunk 3]
```diff
@@ -163,12 +218,21 @@ func DiskSizeBytes(ctx context.Context, inputDir string) (int64, error) {
...
-		total += blockSize + info.Sys().(*syscall.Stat_t).Blocks*512
+		// Estimate from the entry type and size rather than from the blocks
+		// the host filesystem allocated, which vary with the filesystem type
+		// and allocation state and would make the image size depend on the
+		// host. Count a block per entry for its inode and metadata, plus the
+		// data blocks a regular file needs and a data block for each
+		// directory.
+		total += blockSize
+		switch {
+		case info.Mode().IsRegular():
```
Comment (maggie-lou) [inline enterprise/server/util/ext4/ext4.go:201]: One edge case my agent highlighted is sparse files: "sizes sparse files by logical size, so a tiny 1 TB sparse file produces a ~1.2 TB image"

A proposed fix: "While unpacking the image, calculate how much space the files will need based on the image archive itself, after applying overwrites and deletions. This is better because it is deterministic without depending on the host filesystem, while still preserving sparse files instead of treating their empty regions as real data like the PR does"

File: enterprise/server/util/ext4/ext4.go [hunk 4]
```diff
@@ -77,10 +88,31 @@ func DirectoryToImage(ctx context.Context, inputDir, outputFile string, sizeByte
...
+		// - Clamp extraction-time atimes and ctimes to the newest non-future
+		//   file mtime (SOURCE_DATE_EPOCH)
+		// - Skip xattrs, which may contain host-assigned security.selinux
+		//   labels on SELinux hosts (-E no_copy_xattrs)
+		//
+		// mke2fs only clamps timestamps that are newer than the epoch, so the
+		// epoch must not be in the future. Bazel install dirs deliberately set
+		// file mtimes ten years ahead, so those files are skipped when picking
+		// the newest mtime.
+		epoch, err := newestFileModTime(inputDir, time.Now())
```
Comment (maggie-lou) [inline enterprise/server/util/ext4/ext4.go]: This feels like it could be flaky. Could we set SOURCE_DATE_EPOCH to a more deterministic value? If the image changes, the digest should change, so it seems like we shouldn't need to pick up diffs related to mtime updates

File: enterprise/server/util/ext4/ext4.go (hunk 3 above)
Comment (maggie-lou) [reply enterprise/server/util/ext4/ext4.go:201]: A sparse file would make a bug around https://github.com/buildbuddy-io/buildbuddy/pull/13430#discussion_r4065518857 more likely

**Comment by maggie-lou** [comment]
> would love to flip this on and remove the flag if it works ok

An edge case with https://github.com/buildbuddy-io/buildbuddy/pull/13095 is that until a complete snapshot is fully uploaded and a snapshot manifest is published that other executors can use, concurrent executors may try to convert and upload a chunked EXT4 at the same time. The generated EXT4s will not be deterministic, so the chunks will not be deduplicated by the cache and each executor might upload a complete image to the cache, generating a lot of unnecessary cache writes ("image ray of death" issues) This PR/flag would help resolve that

File: enterprise/server/util/ext4/ext4.go (hunk 4 above)
Comment (maggie-lou) [reply enterprise/server/util/ext4/ext4.go]: If fixing the disk size is too complex for now, we can punt on that, but this timestamp diff seems more likely to cause issues and worth addressing

File: enterprise/server/util/ext4/ext4.go (hunk 3 above)
Comment (bduffany) [reply enterprise/server/util/ext4/ext4.go:201]: The current callers of this func don't support sparse files:
- When extracting OCI images (before converting to EXT4), the library we are using doesn't support sparse files, so if someone tried to convert an image with a 1TB file, the image conversion would probably time out
- We also don't support sparse files as workspace inputs

If we want to support sparse files I think we need more comprehensive changes, which we could do independently of this change.

File: enterprise/server/util/ext4/ext4.go (hunk 4 above)
Comment (bduffany) [reply enterprise/server/util/ext4/ext4.go]: Done. This was handling an edge case (pre-extracted bazel installations, which use `mtime` to detect a corrupted install), but I don't think it's worth handling that case for now. I think all of our tests use bazelisk instead of the pre-extracted bazel install.

Comment (maggie-lou) [reply enterprise/server/util/ext4/ext4.go]: I see. So shouldn't be an issue when this is enabled for our tests-only, but it has the potential to break customer images if we enable the flag more broadly

Was looking for whether we pre-extract bazel installations anywhere, and found this [case](https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/dockerfiles/rbe-ubuntu20-04-workflows/Dockerfile#L9). I don't know if it's worth invalidating the image for all users, but pre-extracting bazel 6 is probably unnecessary...

**Review by maggie-lou — APPROVED** [review]
